package/utils/uencrypt/src/uencrypt-mbedtls.c

```diff
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later
+ * Copyright (C) 2023 Eneas Ulir de Queiroz
+ */
+
+#include <ctype.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <mbedtls/cipher.h>
+
+int do_crypt(FILE *infile, FILE *outfile, const mbedtls_cipher_info_t *cipher_info,
+	     const unsigned char *key, const unsigned char *iv, int enc, int padding)
+{
+    mbedtls_cipher_context_t ctx;
+    unsigned char inbuf[1024], outbuf[1024 + MBEDTLS_MAX_BLOCK_LENGTH];
+    size_t inlen, outlen;
+    int ret = 0;
+    int step;
+
+    mbedtls_cipher_init(&ctx);
+    if ((ret = mbedtls_cipher_setup(&ctx, cipher_info))) {
+	fprintf(stderr, "Error: mbedtls_cipher_setup: %d\n", ret);
+	goto out;
+    }
+    step = iv ? 1024 : mbedtls_cipher_get_block_size(&ctx);
+    if ((ret = mbedtls_cipher_setkey(&ctx, key, (int) mbedtls_cipher_get_key_bitlen(&ctx),
+			      enc ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT))) {
+	fprintf(stderr, "Error: mbedtls_cipher_setkey: %d\n", ret);
+	goto out;
+    }
+    if (iv && (ret = mbedtls_cipher_set_iv(&ctx, iv, 16))) {
+	fprintf(stderr, "Error: mbedtls_cipher_set_iv: %d\n", ret);
+	goto out;
+    }
+
+    if (cipher_info->block_size > 1) {
+	if (cipher_info->mode == MBEDTLS_MODE_CBC) {
+	    if ((ret = mbedtls_cipher_set_padding_mode(&ctx,
+						       padding ? MBEDTLS_PADDING_PKCS7
+							       : MBEDTLS_PADDING_NONE))) {
+		fprintf(stderr, "Error: mbedtls_cipher_set_padding_mode: %d\n", ret);
+		goto out;
+	    }
+	} else {
+	    if (cipher_info->mode != MBEDTLS_MODE_CBC && padding) {
+		fprintf(stderr, "Error: mbedTLS only supports padding with CBC ciphers.\n");
+		goto out;
+	    }
+	}
+    }
+
+    if ((ret = mbedtls_cipher_reset(&ctx))) {
+	fprintf(stderr, "Error: mbedtls_cipher_reset: %d\n", ret);
+	goto out;
+    }
+
+    for (;;) {
+	inlen = fread(inbuf, 1, step, infile);
```

Comment:
If `iv` is not set `step = mbedtls_cipher_get_block_size(&ctx);`.
If think you should make sure that inbuf is at least `mbedtls_cipher_get_block_size()` long. For most crypto algorithms  I know this is true.

Reply:
Good idea.  Right now the max is 16 for anything ECB.  A cipher with a large block size may later be added to the library.

I will check the value at compile-time, adjusting the buffer size if it falls under the library's maximum block length--currently 16 in mbedtls, 32 in openssl (because of XTS).

I'll make a run-time check, just in case a new version of the library introduces a cipher with block size>1024, while keeping ABI compatibility with an earlier version with a smaller value.

Should we warn--or even bail out--if that condition ever gets true?  It should fail the next operation, just as it would if it were not ECB, so I'll just set it to the buffer size and let it proceed.